File: Structural/Facade/CSharp/HomeTheaterFacade.cs
// HomeTheaterFacade.cs

using System;

public class HomeTheaterFacade {
  private DVDPlayer dvdPlayer;
  private Projector projector;
  private SoundSystem soundSystem;
  private Screen screen;

  public HomeTheaterFacade(DVDPlayer dvdPlayer, Projector projector, SoundSystem soundSystem, Screen screen) {
    this.dvdPlayer = dvdPlayer;
    this.projector = projector;
    this.soundSystem = soundSystem;
    this.screen = screen;
  }

  public void WatchMovie(string movie) {
    Console.WriteLine("\nGetting ready to watch a movie...");
    screen.Lower();
    projector.TurnOn();
    projector.SetInput("DVD");
    soundSystem.TurnOn();
    soundSystem.SetVolume(10);
    dvdPlayer.TurnOn();
    dvdPlayer.Play(movie);
  }

  public void EndMovie() {
    Console.WriteLine("\nShutting down home theater...");
    dvdPlayer.TurnOff();
    soundSystem.TurnOff();
    projector.TurnOff();
    screen.Raise();
  }
}

File: Structural/Facade/Cpp/HomeTheaterFacade.cpp
// HomeTheaterFacade.cpp

#include "HomeTheaterFacade.h"
#include <iostream>

HomeTheaterFacade::HomeTheaterFacade(DVDPlayer dvd, Projector proj, SoundSystem sound, Screen scr)
    : dvdPlayer(dvd), projector(proj), soundSystem(sound), screen(scr) {}

void HomeTheaterFacade::watchMovie(const std::string& movie) {
  std::cout << "\nGetting ready to watch a movie..." << std::endl;
  screen.lower();
  projector.turnOn();
  projector.setInput("DVD");
  soundSystem.turnOn();
  soundSystem.setVolume(10);
  dvdPlayer.turnOn();
  dvdPlayer.play(movie);
}

void HomeTheaterFacade::endMovie() {
  std::cout << "\nShutting down home theater..." << std::endl;
  dvdPlayer.turnOff();
  soundSystem.turnOff();
  projector.turnOff();
  screen.raise();
}
